File: src/routers/categories.py
from typing import List, Optional
from fastapi import APIRouter, Depends, HTTPException, Query, Request
from sqlalchemy.orm import Session

from database import get_session
from services.category_service import CategoryService
from schemas.category import CategoryCreate, CategoryUpdate, CategoryMove, CategoryOut
from schemas.tree import TreeResponse
from utils.filters import parse_param_filters

router = APIRouter(prefix="/categories", tags=["categories"])


def get_service(session: Session = Depends(get_session)) -> CategoryService:
    return CategoryService(session)


@router.get("/tree", response_model=TreeResponse)
def get_tree(
    request: Request,
    start_id: Optional[int] = Query(None),
    service: CategoryService = Depends(get_service),
):
    param_filters = parse_param_filters(dict(request.query_params)) if request else []
    items = service.get_tree(start_id=start_id, param_filters=param_filters)
    return TreeResponse(items=items)


@router.get("/", response_model=List[CategoryOut])
def list_categories(service: CategoryService = Depends(get_service)):
    return service.get_all()


@router.get("/{category_id}", response_model=CategoryOut)
def get_category(category_id: int, service: CategoryService = Depends(get_service)):
    try:
        return service.get_by_id(category_id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.get("/{category_id}/children", response_model=List[CategoryOut])
def get_children(category_id: int, service: CategoryService = Depends(get_service)):
    try:
        return service.get_children(category_id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.get("/{category_id}/descendants", response_model=List[CategoryOut])
def get_descendants(category_id: int, service: CategoryService = Depends(get_service)):
    try:
        return service.get_descendants(category_id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.get("/{category_id}/parents", response_model=List[CategoryOut])
def get_parents(category_id: int, service: CategoryService = Depends(get_service)):
    try:
        return service.get_parents(category_id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.post("/", response_model=CategoryOut, status_code=201)
def create_category(body: CategoryCreate, service: CategoryService = Depends(get_service)):
    try:
        return service.create(name=body.name, parent_id=body.parent_id)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.patch("/{category_id}", response_model=CategoryOut)
def rename_category(category_id: int, body: CategoryUpdate, service: CategoryService = Depends(get_service)):
    try:
        return service.rename(category_id, name=body.name)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.patch("/{category_id}/move", response_model=CategoryOut)
def move_category(category_id: int, body: CategoryMove, service: CategoryService = Depends(get_service)):
    try:
        return service.move(category_id, new_parent_id=body.new_parent_id)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.delete("/{category_id}", status_code=204)
def delete_category(
    category_id: int,
    cascade: bool = False,
    service: CategoryService = Depends(get_service),
):
    try:
        service.delete(category_id, cascade=cascade)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

File: frontend/src/api/categories.ts
import client from './client';
import type { Category } from '../types';

export type TreeCatItem = { type: 'category'; id: number; name: string; level: number; parent_id: number | null };
export type TreePosItem = {
  type: 'position'; id: number; name: string; category_id: number; level: number;
  parameters: Array<{ parameter_id: number; short_name: string | null; val_real: number | null; val_int: number | null; val_str: string | null; val_dt: string | null; enum_val_id: number | null }>;
};
export type TreeResponse = { items: Array<TreeCatItem | TreePosItem> };

export const categoriesApi = {
  getAll: () => client.get<Category[]>('/categories').then(r => r.data),

  /** GET /categories/tree — returns categories+positions with levels, supports param filters */
  getTree: (params?: Record<string, string | number>) =>
    client.get<TreeResponse>('/categories/tree', { params }).then(r => r.data),

  getById: (id: number) =>
    client.get<Category>(`/categories/${id}`).then(r => r.data),

  getChildren: (id: number) =>
    client.get<Category[]>(`/categories/${id}/children`).then(r => r.data),

  getDescendants: (id: number) =>
    client.get<Category[]>(`/categories/${id}/descendants`).then(r => r.data),

  /** GET /categories/{id}/parents — returns ancestor chain root→parent */
  getParents: (id: number) =>
    client.get<{ id: number; name: string; parent_id: number | null }[]>(`/categories/${id}/parents`).then(r => r.data),

  create: (data: { name: string; parent_id: number | null }) =>
    client.post<Category>('/categories', data).then(r => r.data),

  update: (id: number, data: { name: string }) =>
    client.patch<Category>(`/categories/${id}`, data).then(r => r.data),

  move: (id: number, new_parent_id: number | null) =>
    client.patch<Category>(`/categories/${id}/move`, { new_parent_id }).then(r => r.data),

  delete: (id: number, cascade = false) =>
    client.delete(`/categories/${id}`, { params: { cascade } }),
};
